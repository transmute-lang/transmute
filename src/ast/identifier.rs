use crate::lexer::Span;
use std::fmt::{Display, Formatter};

#[derive(Debug, PartialEq)]
pub struct Identifier {
    name: String,
    span: Span,
}

impl Identifier {
    pub fn new(name: String, span: Span) -> Self {
        Self { name, span }
    }

    pub fn span(&self) -> &Span {
        &self.span
    }

    pub fn name(&self) -> &str {
        &self.name
    }
}

impl Display for Identifier {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.name)
    }
}
