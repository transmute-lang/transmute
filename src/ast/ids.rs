macro_rules! make_id {
    ($name:ident) => {
        #[derive(Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash)]
        pub struct $name {
            id: usize,
        }

        impl $name {
            pub fn id(&self) -> usize {
                self.id
            }
        }

        impl std::fmt::Display for $name {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                write!(f, "{}", self.id)
            }
        }

        impl From<usize> for $name {
            fn from(id: usize) -> Self {
                Self { id }
            }
        }
    };
}

pub(crate) use make_id;

make_id!(IdentId);
make_id!(IdentRefId);
make_id!(ExprId);
make_id!(StmtId);
make_id!(SymbolId);
