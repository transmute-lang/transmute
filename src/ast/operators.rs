use crate::lexer::{Location, Span};

#[derive(Debug, PartialEq)]
pub struct BinaryOperator {
    kind: BinaryOperatorKind,
    location: Location,
    span: Span,
}

impl BinaryOperator {
    pub fn new(kind: BinaryOperatorKind, location: Location, span: Span) -> Self {
        Self {
            kind,
            location,
            span,
        }
    }

    pub fn kind(&self) -> &BinaryOperatorKind {
        &self.kind
    }
}

#[derive(Debug, PartialEq)]
pub enum BinaryOperatorKind {
    Addition,
}
