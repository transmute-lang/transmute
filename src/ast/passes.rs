pub mod exit_points_resolver;
pub mod implicit_ret_converter;
pub mod operators_converter;
