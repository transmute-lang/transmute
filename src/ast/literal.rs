use crate::ast::ids::IdentRefId;
use crate::lexer::Span;
use std::fmt::{Debug, Formatter};

#[derive(Debug, Clone, PartialEq)]
pub struct Literal {
    kind: LiteralKind,
    span: Span,
}

impl Literal {
    pub fn new(kind: LiteralKind, span: Span) -> Self {
        Self { kind, span }
    }

    pub fn kind(&self) -> &LiteralKind {
        &self.kind
    }

    pub fn span(&self) -> &Span {
        &self.span
    }
}

#[derive(Clone, PartialEq)]
pub enum LiteralKind {
    Boolean(bool),
    Identifier(IdentRefId),
    Number(i64),
}

impl Debug for LiteralKind {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            LiteralKind::Boolean(v) => write!(f, "Boolean({v})"),
            LiteralKind::Identifier(v) => write!(f, "IdentRefId({v})"),
            LiteralKind::Number(v) => write!(f, "Number({v})"),
        }
    }
}
