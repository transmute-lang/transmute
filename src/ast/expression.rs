use crate::ast::literal::Literal;
use crate::ast::operators::{BinaryOperator, UnaryOperator};
use crate::ast::Visitor;
use crate::lexer::{Location, Span};

#[derive(Debug, PartialEq)]
pub struct Expression {
    kind: ExpressionKind,
}

impl Expression {
    pub fn new(kind: ExpressionKind) -> Self {
        Self { kind }
    }

    pub fn kind(&self) -> &ExpressionKind {
        &self.kind
    }

    pub fn location(&self) -> &Location {
        match &self.kind {
            ExpressionKind::Literal(l) => l.location(),
            ExpressionKind::Binary(e, _, _) => e.location(),
            ExpressionKind::Unary(o, _) => o.location(),
        }
    }

    pub fn span(&self) -> Span {
        match &self.kind {
            ExpressionKind::Literal(l) => l.span().clone(),
            ExpressionKind::Binary(l, _, r) => l.span().extend_to(&r.span()),
            ExpressionKind::Unary(o, e) => o.span().extend_to(&e.span()),
        }
    }

    pub fn accept<V, R>(&self, visitor: &mut V) -> R
    where
        V: Visitor<R>,
    {
        visitor.visit_expression(self)
    }
}

impl From<Literal> for Expression {
    fn from(literal: Literal) -> Self {
        Self {
            kind: ExpressionKind::Literal(literal),
        }
    }
}

#[derive(Debug, PartialEq)]
pub enum ExpressionKind {
    Literal(Literal),
    Binary(Box<Expression>, BinaryOperator, Box<Expression>),
    Unary(UnaryOperator, Box<Expression>),
}
