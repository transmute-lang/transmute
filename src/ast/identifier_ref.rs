use crate::ast::identifier::Identifier;
use crate::ast::ids::{IdentId, IdentRefId, SymbolId};
use crate::lexer::Span;

/// Represents an identifier when used as a reference
#[derive(Debug, Clone, PartialEq)]
pub struct IdentifierRef<B>
where
    B: BoundState,
{
    /// ID of this identifier reference
    id: IdentRefId,
    /// The referenced symbol identifier
    ident: Identifier,
    /// When `Bound`, the referenced `SymbolId`
    bound_state: B, // todo move to Identifier
}

impl<S> IdentifierRef<S>
where
    S: BoundState,
{
    pub fn id(&self) -> IdentRefId {
        self.id
    }

    pub fn ident(&self) -> &Identifier {
        &self.ident
    }

    pub fn ident_id(&self) -> IdentId {
        self.ident.id()
    }

    pub fn span(&self) -> &Span {
        self.ident.span()
    }
}

impl IdentifierRef<Unbound> {
    pub fn new(id: IdentRefId, ident: Identifier) -> Self {
        Self {
            id,
            ident,
            bound_state: Unbound {},
        }
    }

    pub fn resolved(self, symbol_id: SymbolId) -> IdentifierRef<Bound> {
        IdentifierRef {
            id: self.id,
            ident: self.ident,
            bound_state: Bound(symbol_id),
        }
    }
}

impl IdentifierRef<Bound> {
    pub fn new_resolved(id: IdentRefId, ident: Identifier, symbol_id: SymbolId) -> Self {
        Self {
            id,
            ident,
            bound_state: Bound(symbol_id),
        }
    }
    pub fn symbol_id(&self) -> SymbolId {
        self.bound_state.0
    }
}

pub trait BoundState {}

#[derive(Debug, Clone, PartialEq)]
pub struct Unbound;

impl BoundState for Unbound {}

#[derive(Debug, Clone, PartialEq)]
pub struct Bound(pub SymbolId);

impl BoundState for Bound {}
