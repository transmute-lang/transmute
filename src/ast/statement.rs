use crate::ast::expression::ExprId;
use crate::ast::identifier::Identifier;
use crate::lexer::Span;
use std::fmt::{Display, Formatter};

#[derive(Debug, PartialEq)]
pub struct Statement<'s> {
    id: StmtId,
    // todo should be merged with Expression, maybe?
    kind: StatementKind<'s>,
    span: Span,
}

impl<'s> Statement<'s> {
    pub fn new(id: StmtId, kind: StatementKind<'s>, span: Span) -> Self {
        Self { id, kind, span }
    }

    pub fn id(&self) -> StmtId {
        self.id
    }

    pub fn kind(&self) -> &StatementKind<'s> {
        &self.kind
    }
}

#[derive(Debug, Clone, Copy, Eq, PartialEq)]
pub struct StmtId {
    id: usize,
}

impl StmtId {
    pub fn id(&self) -> usize {
        self.id
    }
}

impl Display for StmtId {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.id)
    }
}

impl From<usize> for StmtId {
    fn from(id: usize) -> Self {
        Self { id }
    }
}

#[derive(Debug, PartialEq)]
pub enum StatementKind<'s> {
    Expression(ExprId),
    Let(Identifier<'s>, ExprId),
    Ret(ExprId),
    // todo vec does not hold span and position as it should (search for "Vec<Statement>, Span")
    LetFn(Identifier<'s>, Vec<Identifier<'s>>, Vec<StmtId>),
}
