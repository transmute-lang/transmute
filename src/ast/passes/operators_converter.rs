use crate::ast::expression::{Expression, ExpressionKind, Untyped};
use crate::ast::identifier::Identifier;
use crate::ast::identifier_ref::{IdentifierRef, Unbound};
use crate::ast::ids::{ExprId, IdentId, IdentRefId};
use crate::vec_map::VecMap;
use bimap::BiMap;

pub struct OperatorsConverter {
    identifiers: BiMap<IdentId, String>,
    identifier_refs: VecMap<IdentRefId, IdentifierRef<Unbound>>,
}

pub struct Output {
    pub identifiers: VecMap<IdentId, String>,
    pub identifier_refs: VecMap<IdentRefId, IdentifierRef<Unbound>>,
    pub expressions: VecMap<ExprId, Expression<Untyped>>,
}

impl OperatorsConverter {
    pub fn new(
        identifiers: VecMap<IdentId, String>,
        identifier_refs: VecMap<IdentRefId, IdentifierRef<Unbound>>,
    ) -> Self {
        Self {
            identifiers: identifiers.into_iter().collect::<BiMap<IdentId, String>>(),
            identifier_refs,
        }
    }

    pub fn convert(mut self, expressions: VecMap<ExprId, Expression<Untyped>>) -> Output {
        let expressions = expressions
            .into_iter()
            .map(|(expr_ir, expr)| (expr_ir, self.convert_expression(expr)))
            .collect::<VecMap<ExprId, Expression<Untyped>>>();

        let identifiers = self
            .identifiers
            .into_iter()
            .collect::<VecMap<IdentId, String>>();

        Output {
            identifiers,
            identifier_refs: self.identifier_refs,
            expressions,
        }
    }

    fn convert_expression(&mut self, expression: Expression<Untyped>) -> Expression<Untyped> {
        match expression.kind() {
            ExpressionKind::Binary(lhs_id, op, rhs_id) => {
                let ident_id = self.ident_id(op.kind().function_name());
                let identifier = Identifier::new(ident_id, op.span().clone());
                let ident_ref_id = IdentRefId::from(self.identifier_refs.len());
                self.identifier_refs
                    .push(IdentifierRef::<Unbound>::new(ident_ref_id, identifier));

                Expression::new(
                    expression.id(),
                    ExpressionKind::FunctionCall(ident_ref_id, vec![*lhs_id, *rhs_id]),
                    expression.span().clone(),
                )
            }
            ExpressionKind::Unary(op, expr_id) => {
                let ident_id = self.ident_id(op.kind().function_name());
                let identifier = Identifier::new(ident_id, op.span().clone());
                let ident_ref_id = IdentRefId::from(self.identifier_refs.len());
                self.identifier_refs
                    .push(IdentifierRef::<Unbound>::new(ident_ref_id, identifier));

                Expression::new(
                    expression.id(),
                    ExpressionKind::FunctionCall(ident_ref_id, vec![*expr_id]),
                    expression.span().clone(),
                )
            }
            _ => expression,
        }
    }

    fn ident_id(&mut self, function_name: &'static str) -> IdentId {
        if !self.identifiers.contains_right(function_name) {
            let ident_id = IdentId::from(self.identifiers.len());
            self.identifiers.insert(ident_id, function_name.to_string());
        }

        *self
            .identifiers
            .get_by_right(function_name)
            .expect("it is there")
    }
}

#[cfg(test)]
mod tests {
    use insta::{assert_snapshot};
    use crate::lexer::Lexer;
    use crate::output::pretty_print::Options;
    use crate::parser::Parser;

    macro_rules! op {
        ($name: ident, $src:expr) => {
            #[test]
            fn $name() {
                let ast = Parser::new(Lexer::new($src))
                .parse()
                .unwrap()
                .convert_operators();

                let mut w = String::new();
                ast.pretty_print(&Options::default(), &mut w).unwrap();

                assert_snapshot!(w);
            }
        };
    }
    
    op!(add, "lhs + rhs;");
    op!(sub, "lhs - rhs;");
    op!(mul, "lhs * rhs;");
    op!(div, "lhs / rhs;");
    op!(eq, "lhs == rhs;");
    op!(neq, "lhs != rhs;");
    op!(gt, "lhs > rhs;");
    op!(ge, "lhs >= rhs;");
    op!(lt, "lhs < rhs;");
    op!(le, "lhs <= rhs;");

    op!(minus, "- op;");
}
