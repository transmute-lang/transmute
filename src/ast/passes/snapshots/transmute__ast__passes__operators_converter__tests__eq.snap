---
source: src/ast/passes/operators_converter.rs
expression: w
---
eq(lhs, rhs);

